programs/drift/src/controller/liquidation.rs

```diff
@@ -648,6 +654,464 @@ pub fn liquidate_perp(
+    liquidator_loader: &AccountLoader<User>,
+    liquidator_key: &Pubkey,
+    liquidator_stats_loader: &AccountLoader<UserStats>,
+    makers_and_referrer: &UserMap,
+    makers_and_referrer_stats: &UserStatsMap,
+    perp_market_map: &PerpMarketMap,
+    spot_market_map: &SpotMarketMap,
+    oracle_map: &mut OracleMap,
+    clock: &Clock,
+    state: &State,
+) -> DriftResult {
+    let now = clock.unix_timestamp;
+    let slot = clock.slot;
+
+    let mut user = load_mut!(user_loader)?;
+    let mut liquidator = load_mut!(liquidator_loader)?;
+
+    let liquidation_margin_buffer_ratio = state.liquidation_margin_buffer_ratio;
+    let initial_pct_to_liquidate = state.initial_pct_to_liquidate as u128;
+    let liquidation_duration = state.liquidation_duration as u128;
+
+    validate!(
+        !user.is_bankrupt(),
+        ErrorCode::UserBankrupt,
+        "user bankrupt",
+    )?;
+
+    validate!(
+        !liquidator.is_bankrupt(),
+        ErrorCode::UserBankrupt,
+        "liquidator bankrupt",
+    )?;
+
+    let market = perp_market_map.get_ref(&market_index)?;
+
+    validate!(
+        !market.is_operation_paused(PerpOperation::Liquidation),
+        ErrorCode::InvalidLiquidation,
+        "Liquidation operation is paused for market {}",
+        market_index
+    )?;
+
+    drop(market);
+
+    // Settle user's funding payments so that collateral is up to date
+    settle_funding_payment(
+        &mut user,
+        user_key,
+        perp_market_map.get_ref_mut(&market_index)?.deref_mut(),
+        now,
+    )?;
+
+    // Settle user's funding payments so that collateral is up to date
+    settle_funding_payment(
+        &mut liquidator,
+        liquidator_key,
+        perp_market_map.get_ref_mut(&market_index)?.deref_mut(),
+        now,
+    )?;
+
+    let margin_calculation = calculate_margin_requirement_and_total_collateral_and_liability_info(
+        &user,
+        perp_market_map,
+        spot_market_map,
+        oracle_map,
+        MarginContext::liquidation(liquidation_margin_buffer_ratio)
+            .track_market_margin_requirement(MarketIdentifier::perp(market_index))?,
+    )?;
+
+    if !user.is_being_liquidated() && margin_calculation.meets_margin_requirement() {
+        msg!("margin calculation: {:?}", margin_calculation);
+        return Err(ErrorCode::SufficientCollateral);
+    } else if user.is_being_liquidated() && margin_calculation.can_exit_liquidation()? {
+        user.exit_liquidation();
+        return Ok(());
+    }
+
+    user.get_perp_position(market_index).map_err(|e| {
+        msg!(
+            "User does not have a position for perp market {}",
+            market_index
+        );
+        e
+    })?;
+
+    liquidator
+        .force_get_perp_position_mut(market_index)
+        .map_err(|e| {
+            msg!(
+                "Liquidator has no available positions to take on perp position in market {}",
+                market_index
+            );
+            e
+        })?;
```

Comment:
more on here, is it necessary to enforce this no position liquidator have available position slot? similar to how ref reward is best-effort, is that worthwhile here?

Reply:
i think we can remove this and the liquidator will just get no filler reward